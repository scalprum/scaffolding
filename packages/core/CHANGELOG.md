# Changelog

This file was generated using [@jscutlery/semver](https://github.com/jscutlery/semver).

## [0.8.3](https://github.com/scalprum/scaffloding/compare/@scalprum/core-0.8.2...@scalprum/core-0.8.3) (2025-03-31)


### Bug Fixes

* **core:** do not process manifest for intialized scopes ([dcd4bed](https://github.com/scalprum/scaffloding/commit/dcd4bedeed50f9e17c45fbf46eccd4e270de0771))

## [0.8.2](https://github.com/scalprum/scaffloding/compare/@scalprum/core-0.8.1...@scalprum/core-0.8.2) (2025-03-20)

## [0.8.1](https://github.com/scalprum/scaffolding/compare/@scalprum/core-0.8.0...@scalprum/core-0.8.1) (2024-09-11)


### Bug Fixes

* **core:** ensure scalprum instance receives configuration updates ([9caf092](https://github.com/scalprum/scaffolding/commit/9caf092b741300cfd395b42844e21804204a297c))

## [0.8.0](https://github.com/scalprum/scaffolding/compare/@scalprum/core-0.7.0...@scalprum/core-0.8.0) (2024-09-09)


### Features

* **core:** allow directly using plugin manifest ([9eede15](https://github.com/scalprum/scaffolding/commit/9eede15da2db3113f480326597f612e8cd853840))

## [0.7.0](https://github.com/scalprum/scaffolding/compare/@scalprum/core-0.6.6...@scalprum/core-0.7.0) (2024-01-22)


### Features

* support SDK v5 ([aa922b7](https://github.com/scalprum/scaffolding/commit/aa922b710d50c2ae5058a4b11a623c93ce89edcf))

## [0.6.6](https://github.com/scalprum/scaffolding/compare/@scalprum/core-0.6.5...@scalprum/core-0.6.6) (2024-01-22)

## [0.6.5](https://github.com/scalprum/scaffolding/compare/@scalprum/core-0.6.4...@scalprum/core-0.6.5) (2023-12-04)

## [0.6.4](https://github.com/scalprum/scaffolding/compare/@scalprum/core-0.6.3...@scalprum/core-0.6.4) (2023-12-04)

## [0.6.3](https://github.com/scalprum/scaffolding/compare/@scalprum/core-0.6.2...@scalprum/core-0.6.3) (2023-12-04)

# Changelog
