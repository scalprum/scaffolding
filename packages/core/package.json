{
  "name": "@scalprum/core",
  "version": "0.5.1",
  "description": "Includes core functions for scalprum scaffolding.",
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "files": [
    "dist"
  ],
  "typings": "dist/index.d.ts",
  "repository": "https://github.com/scalprum/scaffloding.git",
  "author": "Martin Marosi <marvusm.mmi@gmail.com>",
  "license": "Apache-2.0",
  "scripts": {
    "build": "npm run build:cjs && npm run build:esm && npm run build:typings",
    "build:esm": "tsc --module es2015 --target es5",
    "build:cjs": "tsc -p tsconfig.cjs.json",
    "build:typings": "node ../../scripts/create-typings.js",
    "start:esm": "npm run build:esm -- -w",
    "start:cjs": "npm run build:cjs -- -w"
  },
  "dependencies": {
    "@openshift/dynamic-plugin-sdk": "file:../../../dynamic-plugin-sdk/packages/lib-core"
  }
}
