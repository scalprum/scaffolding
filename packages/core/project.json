{
  "name": "@scalprum/core",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/core/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@scalprum/build-utils:builder",
      "options": {
        "outputPath": "dist/packages/core",
        "esmTsConfig": "packages/core/tsconfig.esm.json",
        "cjsTsConfig": "packages/core/tsconfig.cjs.json",
        "assets": ["packages/core/*.md"]
      }
    },
    "publish": {
      "command": "node tools/scripts/publish.mjs @scalprum/core {args.ver} {args.tag}",
      "dependsOn": ["build"]
    },
    "lint": {
      "executor": "@nx/eslint:lint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/core/**/*.ts", "packages/core/package.json"]
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "packages/core/jest.config.ts"
      }
    },
    "version": {
      "executor": "@jscutlery/semver:version",
      "options": {
        "push": true,
        "preset": "conventionalcommits",
        "commitMessageFormat": "chore: bump {projectName} to {version} [skip ci]"
      }
    },
    "github": {
      "executor": "@jscutlery/semver:github",
      "options": {
        "tag": "{tag}",
        "notes": "{notes}"
      }
    },
    "npm": {
      "executor": "ngx-deploy-npm:deploy",
      "options": {
        "access": "public",
        "distFolderPath": "dist/packages/core"
      },
      "dependsOn": ["build"]
    },
    "syncDependencies": {
      "executor": "@scalprum/build-utils:sync-dependencies"
    }
  },
  "tags": []
}
