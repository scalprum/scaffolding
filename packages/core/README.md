# @scalprum/core

**Framework-agnostic core for micro-frontend module federation**

The `@scalprum/core` package provides the foundational module federation capabilities for Scalprum. It's a framework-agnostic library that handles dynamic module loading, caching, and manifest processing - making it compatible with any JavaScript framework, not just React.

## Installation

```bash
npm install @scalprum/core
```

## Key Features

- **Framework Agnostic**: Works with any JavaScript framework or vanilla JS
- **Dynamic Module Loading**: Load remote modules at runtime with caching
- **Manifest Processing**: Support for both plugin manifests and custom formats
- **Shared Scope Management**: Integration with webpack's module federation shared scopes
- **Shared Stores**: Event-driven state management for micro-frontends
- **Built-in Caching**: Intelligent module caching with configurable timeout
- **Error Handling**: Robust error handling for network and module loading failures

## Basic Usage

```typescript
import { initialize, getModule, AppsConfig } from '@scalprum/core';

// Configure your remote modules
const config: AppsConfig = {
  myRemoteApp: {
    name: 'myRemoteApp',
    manifestLocation: 'http://localhost:3001/plugin-manifest.json'
  }
};

// Initialize Scalprum
const scalprum = initialize({
  appsConfig: config,
  api: { /* shared context */ }
});

// Load a module dynamically
const MyComponent = await getModule('myRemoteApp', 'MyComponent');
```

## API Reference

### Core Functions

#### `initialize(options)`

Initializes the Scalprum instance with configuration.

**Parameters:**
- `appsConfig: AppsConfig` - Configuration for remote modules
- `api?: any` - Shared API context available to all modules
- `options?: Partial<ScalprumOptions>` - Optional configuration
- `pluginStoreFeatureFlags?: FeatureFlags` - Feature flags for plugin store
- `pluginLoaderOptions?: PluginLoaderOptions` - Options for plugin loader
- `pluginStoreOptions?: PluginStoreOptions` - Options for plugin store

**Returns:** `Scalprum` instance

#### `getModule<T>(scope, module, importName?)`

Loads a module from a remote container.

**Parameters:**
- `scope: string` - The remote container name
- `module: string` - The module name to load
- `importName?: string` - Specific export name (defaults to 'default')

**Returns:** `Promise<T>` - The loaded module

#### `getScalprum()`

Gets the current Scalprum instance.

**Returns:** `Scalprum` instance

**Throws:** Error if Scalprum hasn't been initialized

### Configuration Types

#### `AppsConfig`

```typescript
interface AppsConfig<T = {}> {
  [key: string]: AppMetadata<T>;
}

type AppMetadata<T = {}> = T & {
  name: string;
  appId?: string;
  elementId?: string;
  rootLocation?: string;
  scriptLocation?: string;
  manifestLocation?: string;
  pluginManifest?: PluginManifest;
};
```

#### `ScalprumOptions`

```typescript
interface ScalprumOptions {
  cacheTimeout: number;        // Cache timeout in seconds (default: 120)
  enableScopeWarning: boolean; // Enable duplicate package warnings
}
```

## Advanced Usage

### Preloading Modules

```typescript
import { preloadModule } from '@scalprum/core';

// Preload a module without importing it
await preloadModule('myRemoteApp', 'MyComponent');

// With custom manifest processor
await preloadModule('myRemoteApp', 'MyComponent', (manifest) => manifest.assets.js);

// Later, get the cached module instantly
const MyComponent = await getModule('myRemoteApp', 'MyComponent');
```

### Custom Manifest Processing

```typescript
const processor = (manifest: any) => {
  // Extract entry scripts from custom manifest format
  return manifest.assets.js;
};

await processManifest(
  'http://localhost:3001/custom-manifest.json',
  'myScope',
  'MyModule',
  processor
);
```

### Module Caching

```typescript
import { getCachedModule } from '@scalprum/core';

// Check if module is cached
const { cachedModule, prefetchPromise } = getCachedModule('myScope', 'MyModule');

if (cachedModule) {
  // Module is cached and ready
  const component = cachedModule.default;
}
```

## Shared Scope Integration

Scalprum integrates with webpack's module federation shared scopes:

```typescript
import { initSharedScope, getSharedScope } from '@scalprum/core';

// Initialize shared scope
await initSharedScope();

// Get shared scope object
const sharedScope = getSharedScope(true); // true enables duplicate warnings
```

## Error Handling

```typescript
try {
  const module = await getModule('myScope', 'NonExistentModule');
} catch (error) {
  if (error.message.includes('Module not initialized')) {
    // Module wasn't found in the remote container
    console.error('Module not available:', error);
  } else if (error.message.includes('Manifest location not found')) {
    // Scope configuration is missing manifestLocation
    console.error('Configuration error:', error);
  }
}
```

## Build Tool Compatibility

This package is compatible with:

- **Webpack 5** with Module Federation
- **Rspack** with Module Federation support
- **Module Federation Runtime** for any bundler

## Shared Stores

The `createSharedStore` function enables event-driven state management across microfrontends. See the comprehensive [Shared Stores Guide](../react-core/docs/shared-stores.md) for detailed documentation.

**Quick Example:**

```typescript
import { createSharedStore } from '@scalprum/core';

const EVENTS = ['UPDATE_USER', 'LOGOUT'] as const;

const store = createSharedStore({
  initialState: { user: null, isAuthenticated: false },
  events: EVENTS,
  onEventChange: (state, event, payload) => {
    switch (event) {
      case 'UPDATE_USER':
        return { ...state, user: payload.user, isAuthenticated: true };
      case 'LOGOUT':
        return { user: null, isAuthenticated: false };
      default:
        return state;
    }
  },
});

// Update state
store.updateState('UPDATE_USER', { user: { id: '123', name: 'John' } });

// Subscribe to changes
const unsubscribe = store.subscribeAll(() => {
  console.log('State changed:', store.getState());
});
```

For React integration, use with `useGetState` and `useSubscribeStore` from `@scalprum/react-core`. See [Shared Stores Documentation](../react-core/docs/shared-stores.md) for complete guide.

## Related Packages

- [`@scalprum/react-core`](../react-core) - React bindings and components
- [`@scalprum/build-utils`](../build-utils) - Build tools and NX executors
- [`@scalprum/react-test-utils`](../react-test-utils) - Testing utilities