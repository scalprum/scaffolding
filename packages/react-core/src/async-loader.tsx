import React from 'react';
import { asyncLoader } from '@scalprum/core';

const DefaultErrorComponent: React.ComponentType<any> = () => {
  return <span>Error while loading component!</span>;
};

export function loadComponent(scope: string, module: string, ErrorComponent: React.ComponentType<any> = DefaultErrorComponent) {
  return async (): Promise<{ default: React.ComponentType<any> }> => {
    let Module;
    try {
      Module = await asyncLoader(scope, module);
    } catch (e) {
      console.error(e);
      Module = {
        default: ErrorComponent,
      };
    }

    return Module;
  };
}
