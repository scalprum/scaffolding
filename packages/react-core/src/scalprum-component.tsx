import React, { useEffect, Suspense, useState, ReactNode, useRef, useMemo } from 'react';
import {
  getCachedModule,
  handlePrefetchPromise,
  getAppData,
  processManifest,
  getPendingLoading,
  setPendingLoading,
  getPendingPrefetch,
} from '@scalprum/core';
import isEqual from 'lodash/isEqual';
import { loadComponent } from './async-loader';
import DefaultErrorComponent from './default-error-component';
import { PrefetchProvider } from './prefetch-context';
import { useScalprum } from './use-scalprum';

// eslint-disable-next-line @typescript-eslint/ban-types
export type ScalprumComponentProps<API extends Record<string, any> = {}, Props extends Record<string, any> = {}> = Props & {
  fallback?: NonNullable<React.ReactNode> | null;
  api?: API;
  scope: string;
  module: string;
  ErrorComponent?: React.ReactElement;
  LoadingComponent?: React.ComponentType;
  innerRef?: React.Ref<unknown>;
  processor?: (item: any) => string[];
  skipCache?: boolean;
};

type LoadModuleProps = Omit<ScalprumComponentProps, 'ErrorComponent'>;

const LoadModule: React.ComponentType<LoadModuleProps> = ({
  fallback = 'loading',
  scope,
  module,
  processor,
  innerRef,
  skipCache = false,
  ...props
}) => {
  const { manifestLocation } = getAppData(scope);
  const [prefetchPromise, setPrefetchPromise] = useState<Promise<any>>();
  const { cachedModule } = getCachedModule(scope, module);
  const prefetchID = `${scope}#${module}`;

  const scalprumApi = useScalprum();

  const initComponent = async (): Promise<React.ComponentType<{ ref?: React.Ref<unknown> } & Record<string, any>>> => {
    const { cachedModule, prefetchPromise } = getCachedModule(scope, module);
    if (!skipCache && cachedModule) {
      setPrefetchPromise(prefetchPromise);
      return cachedModule.default;
    }
    const pendingLoading = getPendingLoading(scope, module);
    if (pendingLoading) {
      return pendingLoading.then(() => {
        const { cachedModule } = getCachedModule(scope, module);
        return cachedModule?.default;
      });
    }
    const processPromise = processManifest(manifestLocation || '', scope, module, processor).then(() =>
      loadComponent(scope, module).then(({ component, prefetch }) => {
        if (prefetch) {
          const prefetchResult = getPendingPrefetch(prefetchID) || prefetch(scalprumApi);
          setPrefetchPromise(prefetchResult);
          handlePrefetchPromise(prefetchID, prefetchResult);
        }
        return component;
      })
    );
    setPendingLoading(module, scope, processPromise);
    return processPromise;
  };

  // clear prefetchPromise (from factory)
  const initialPrefetch = useRef(false);
  useEffect(() => {
    if (initialPrefetch.current) {
      setPrefetchPromise(undefined);
    }
    // handle promise prefetch for duplicate module
    if (!prefetchPromise && cachedModule?.prefetch) {
      const prefetchResult = getPendingPrefetch(prefetchID) || cachedModule.prefetch(scalprumApi);
      setPrefetchPromise(prefetchResult);
      handlePrefetchPromise(prefetchID, prefetchResult);
    }
    initialPrefetch.current = true;
  }, [scope, module]);

  const Component = useMemo(() => {
    if (!skipCache && cachedModule?.default) {
      return cachedModule.default;
    }
    return React.lazy(() =>
      initComponent().then((r) => {
        return { default: r };
      })
    );
  }, [scope, module]);
  return (
    <PrefetchProvider prefetchPromise={prefetchPromise}>
      <Suspense fallback={fallback}>{<Component ref={innerRef} {...props} />}</Suspense>
    </PrefetchProvider>
  );
};

interface BaseScalprumComponentState {
  hasError: boolean;
  error?: any;
  errorInfo?: any;
  repairAttempt?: boolean;
}

class BaseScalprumComponent extends React.Component<ScalprumComponentProps, BaseScalprumComponentState> {
  static defaultProps = {
    ErrorComponent: <DefaultErrorComponent />,
  };
  constructor(props: ScalprumComponentProps) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(error: any): BaseScalprumComponentState {
    return { hasError: true, error };
  }

  shouldComponentUpdate(nextProps: ScalprumComponentProps, nextState: BaseScalprumComponentState) {
    if (this.state.hasError !== nextState.hasError) {
      return true;
    }

    return !isEqual(nextProps, this.props) || !isEqual(nextState, this.state);
  }

  componentDidCatch(error: any, errorInfo: React.ErrorInfo) {
    console.error('Scalprum encountered an error!', error?.cause || error.message, error);
    this.setState({ error, errorInfo });
  }

  render(): ReactNode {
    const { ErrorComponent = <DefaultErrorComponent {...this.state} />, ...props } = this.props;

    if (this.state.hasError) {
      return React.cloneElement(ErrorComponent as React.FunctionComponentElement<BaseScalprumComponentState>, { ...this.state });
    }
    return <LoadModule {...props} />;
  }
}

export const ScalprumComponent: React.ComponentType<ScalprumComponentProps> = React.forwardRef((props, ref) => (
  <BaseScalprumComponent {...props} innerRef={ref} />
));
