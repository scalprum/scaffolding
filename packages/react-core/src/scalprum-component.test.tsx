/* eslint-disable @typescript-eslint/ban-ts-comment */
import React, { ComponentType, useEffect } from 'react';
import * as asyncComponent from './async-loader';
import { ScalprumComponent, ScalprumComponentProps } from './scalprum-component';
import { render, cleanup, act, screen } from '@testing-library/react';
import * as ScalprumCore from '@scalprum/core';
import { AppsConfig } from '@scalprum/core';
import TestComponent from './TestComponent';
import { PluginManifest } from '@openshift/dynamic-plugin-sdk';

const ErrorComponent = () => {
  return <h1>Custom error component</h1>;
};

describe('<ScalprumComponent />', () => {
  const testManifest: PluginManifest = {
    extensions: [],
    loadScripts: [],
    name: 'testApp',
    registrationMethod: 'custom',
    version: '1.0.0',
    baseURL: '/',
  };
  const mockInitScalprumConfig: AppsConfig = {
    appOne: {
      name: 'appOne',
      manifestLocation: '/bar.js',
    },
    error: {
      name: 'error',
      manifestLocation: '/bar.js',
    },
    errorSelfRepair: {
      name: 'errorSelfRepair',
      manifestLocation: '/errorSelfRepair.js',
    },
    errorRepairSuccess: {
      name: 'errorRepairSuccess',
      manifestLocation: '/errorRepairSuccess,js',
    },
  };

  const mockInitScalpumConfigManifest: AppsConfig = {
    appOne: {
      name: 'appOne',
      appId: 'appOne',
      rootLocation: '/foo',
      manifestLocation: '/bar.json',
      elementId: 'id',
    },
    error: {
      name: 'error',
      manifestLocation: '/bar.js',
    },
    errorSelfRepair: {
      name: 'errorSelfRepair',
      manifestLocation: '/errorSelfRepair.js',
    },
    errorRepairSuccess: {
      name: 'errorRepairSuccess',
      manifestLocation: '/errorRepairSuccess,js',
    },
  };
  const getAppDataSpy = jest.spyOn(ScalprumCore, 'getAppData').mockReturnValue(mockInitScalprumConfig.appOne);
  const processManifestSpy = jest.spyOn(ScalprumCore, 'processManifest');
  let loadComponentSpy: jest.SpyInstance;

  beforeAll(() => {
    // @ts-ignore
    global.__webpack_share_scopes__ = {
      default: {},
    };
    // @ts-ignore
    global.__webpack_init_sharing__ = () => undefined;
    global.fetch = () =>
      Promise.resolve({
        ok: true,
        json: () =>
          Promise.resolve({
            appOne: {
              entry: [],
            },
            errorSelfRepair: {
              entry: [],
            },
            errorRepairSuccess: {
              entry: [],
            },
          }),
      }) as unknown as Promise<Response>;
  });

  beforeEach(() => {
    const componentPromise = Promise.resolve({ prefetch: undefined, component: TestComponent });
    loadComponentSpy = jest.spyOn(asyncComponent, 'loadComponent').mockReturnValue(componentPromise);
  });

  afterEach(() => {
    cleanup();
    getAppDataSpy.mockClear();
    processManifestSpy.mockClear();
    loadComponentSpy.mockReset();
  });

  test('should retrieve script location', async () => {
    processManifestSpy.mockImplementationOnce(() => Promise.resolve());
    ScalprumCore.initialize({ appsConfig: mockInitScalprumConfig });
    await act(async () => {
      await render(<ScalprumComponent scope="appOne" module="test" />);
    });

    expect(getAppDataSpy).toHaveBeenCalledWith('appOne');
  });

  test('should retrieve manifest location', async () => {
    processManifestSpy.mockImplementationOnce(() => Promise.resolve());
    getAppDataSpy.mockReturnValueOnce(mockInitScalpumConfigManifest.appOne);
    ScalprumCore.initialize({ appsConfig: mockInitScalpumConfigManifest });
    await act(async () => {
      await render(<ScalprumComponent scope="appOne" module="test" />);
    });

    expect(getAppDataSpy).toHaveBeenCalledWith('appOne');
  });

  test('should inject manifest and mount app if it was not initialized before', async () => {
    getAppDataSpy.mockReturnValueOnce(mockInitScalpumConfigManifest.appOne);
    ScalprumCore.initialize({ appsConfig: mockInitScalpumConfigManifest });
    processManifestSpy.mockImplementationOnce(() => {
      return Promise.resolve();
    });
    await act(async () => {
      render(<ScalprumComponent scope="appOne" module="test" />);
    });

    expect(processManifestSpy).toHaveBeenCalledWith('/bar.json', 'appOne', 'test', undefined);
  });

  test('should render test component', async () => {
    processManifestSpy.mockImplementationOnce(() => Promise.resolve());
    ScalprumCore.initialize({ appsConfig: mockInitScalprumConfig });
    let container;

    const props: ScalprumComponentProps<{ apiProp: () => void }, { testProp: number }> = {
      testProp: 5,
      scope: 'appOne',
      module: 'test',
      api: {
        apiProp: () => undefined,
      },
    };
    await act(async () => {
      container = render(<ScalprumComponent {...props} />)?.container;
      expect(container).toMatchSnapshot();
    });

    expect(loadComponentSpy).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });

  test('should render test component with manifest', async () => {
    jest.useFakeTimers();
    getAppDataSpy.mockReturnValueOnce(mockInitScalpumConfigManifest.appOne);
    ScalprumCore.initialize({ appsConfig: mockInitScalprumConfig });
    processManifestSpy.mockImplementationOnce(() => {
      return Promise.resolve();
    });
    let container;
    await act(async () => {
      container = render(<ScalprumComponent scope="appOne" module="test" />)?.container;
    });

    expect(loadComponentSpy).toHaveBeenCalled();

    /**
     * Wait for the loading promise to be resolved
     */
    await act(async () => {
      jest.advanceTimersByTime(1);
    });

    expect(container).toMatchSnapshot();
  });

  test('should render fallback component', async () => {
    jest.spyOn(global, 'fetch').mockImplementationOnce(
      () =>
        new Promise((res) => {
          setTimeout(() => {
            res({
              ok: true,
              json: () =>
                Promise.resolve({
                  appOne: {
                    entry: [],
                  },
                }),
            } as unknown as Promise<Response>);
          }, 500);
        })
    );
    jest.useFakeTimers();
    /**
     * We need the async component "hang" to render the fallback
     */

    const componentPromise = new Promise<{ prefetch: (() => Promise<any>) | undefined; component: ComponentType<any> }>((res) =>
      setTimeout(() => res({ prefetch: undefined, component: TestComponent }), 500)
    );
    jest.spyOn(asyncComponent, 'loadComponent').mockReturnValueOnce(componentPromise);
    ScalprumCore.initialize({ appsConfig: mockInitScalprumConfig });

    const props: ScalprumComponentProps = {
      scope: 'appOne',
      module: 'test',
      LoadingComponent: () => <h1>Loading</h1>,
      fallback: <h1>Suspense fallback</h1>,
    };
    const { container } = render(<ScalprumComponent {...props} />);
    /**
     * Should render Loading component
     */
    expect(container).toMatchSnapshot();
    await act(async () => {
      jest.advanceTimersByTime(200);
    });
    /**
     * Should render fallback component
     */
    expect(container).toMatchSnapshot();
  });

  test('should render error component', async () => {
    processManifestSpy.mockImplementation(() => Promise.resolve());
    const componentPromise = Promise.resolve({ prefetch: undefined, component: TestComponent });
    const asyncComponentSpy = jest
      .spyOn(asyncComponent, 'loadComponent')
      .mockReturnValueOnce(
        componentPromise.then(() => ({
          prefetch: undefined,
          component: () => {
            throw 'should render error component';
          },
        }))
      )
      .mockReturnValueOnce(
        componentPromise.then(() => ({
          prefetch: undefined,
          component: () => {
            throw 'should render error component';
          },
        }))
      );
    ScalprumCore.initialize({ appsConfig: mockInitScalprumConfig });

    const props: ScalprumComponentProps = {
      scope: 'error',
      module: 'test',
      ErrorComponent: <ErrorComponent />,
    };
    let container;
    await act(async () => {
      container = render(<ScalprumComponent {...props} />).container;
    });
    expect(container).toMatchSnapshot();
    asyncComponentSpy.mockRestore();
  });

  test('should retrieve module from scalprum cache', async () => {
    processManifestSpy.mockImplementation(() => Promise.resolve());
    const cachedModule = {
      __esModule: true,
      default: () => <div data-testid="cached-component">Cached component</div>,
      prefetch: () => Promise.resolve(),
    };
    ScalprumCore.initialize({ appsConfig: mockInitScalprumConfig });
    ScalprumCore.getScalprum().exposedModules[`cachedScope#./test`] = cachedModule;
    await ScalprumCore.getScalprum().pluginStore.loadPlugin(testManifest);

    const props: ScalprumComponentProps = {
      scope: 'cachedScope',
      module: './test',
    };
    let container;
    await act(async () => {
      container = render(<ScalprumComponent {...props} />).container;
    });
    expect(loadComponentSpy).not.toHaveBeenCalled();
    expect(container).toMatchSnapshot();
    expect(screen.getAllByTestId('cached-component')).toHaveLength(1);
  });

  test('should render error component if self-repair attempt fails', async () => {
    // uncomment if you want to see scalprum errors and warnings
    jest.spyOn(console, 'error').mockImplementation(() => undefined);
    jest.spyOn(console, 'warn').mockImplementation(() => undefined);
    processManifestSpy.mockImplementation(() => Promise.reject('Initial fail'));
    const componentPromise = Promise.resolve({ prefetch: undefined, component: TestComponent });
    loadComponentSpy.mockReturnValueOnce(
      componentPromise.then(() => ({
        __esModule: true,
        default: () => {
          useEffect(() => {
            throw new Error('Expected runtime test error');
          }, []);
          return <div>Mocked testing component</div>;
        },
        prefetch: () => Promise.resolve(),
      }))
    );
    ScalprumCore.initialize({ appsConfig: mockInitScalprumConfig });
    let container;
    await act(async () => {
      container = await render(<ScalprumComponent ErrorComponent={<ErrorComponent />} module="./TestComponent" scope="errorSelfRepair" />).container;
    });

    expect(container).toMatchSnapshot();
  });
});
