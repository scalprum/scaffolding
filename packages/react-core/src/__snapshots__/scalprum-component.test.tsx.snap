// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScalprumComponent /> should render error component 1`] = `
<div>
  <h1>
    Custom error component
  </h1>
</div>
`;

exports[`<ScalprumComponent /> should render error component if self-repair attempt fails 1`] = `
<div>
  <h1>
    Custom error component
  </h1>
</div>
`;

exports[`<ScalprumComponent /> should render fallback component 1`] = `
<div>
  <h1>
    Suspense fallback
  </h1>
</div>
`;

exports[`<ScalprumComponent /> should render fallback component 2`] = `
<div>
  <h1>
    Suspense fallback
  </h1>
</div>
`;

exports[`<ScalprumComponent /> should render test component 1`] = `
<div>
  loading
</div>
`;

exports[`<ScalprumComponent /> should render test component 2`] = `
<div>
  <span>
    Test
  </span>
</div>
`;

exports[`<ScalprumComponent /> should render test component with manifest 1`] = `
<div>
  <span>
    Test
  </span>
</div>
`;

exports[`<ScalprumComponent /> should retrieve module from scalprum cache 1`] = `
<div>
  <div
    data-testid="cached-component"
  >
    Cached component
  </div>
</div>
`;
